### PR TITLE
feat(sdk): support more than one exit handler per pipeline

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Enables use of more than one exit handler per pipeline.

**Checklist:**
- [x] The title for your pull request (PR) should follow our 
title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_multiple_exit_handlers.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    first_exit_task = print_op(message='First exit handler has worked!')
+
+    with dsl.ExitHandler(first_exit_task):
+        first_exit_print_task = print_op(message=message)
+        fail_op(message='Task failed.')
+
+    second_exit_task = print_op(message='Second exit handler has worked!')
+
+    with dsl.ExitHandler(second_exit_task):
+        print_op(message=message).after(first_exit_print_task)
```
Comment (chensun) [inline sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_multiple_exit_handlers.py:47]: The after usage is an interesting one. Normally we don't allow dependency reference across DAGs. I wonder if we should disallow this?
In any case, the run result you showed me doesn't reflect the same dependency.

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_multiple_exit_handlers.py:47]: Discussed offline and decided to disallow this to be consistent with dsl condition and looping

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -707,14 +625,10 @@ def _get_dependencies(
...
-                # If a task depends on a condition group or a loop group, it
-                # must explicitly dependent on a task inside the group. This
-                # should not be allowed, because it leads to ambiguous
-                # expectations for runtime behaviors.
+                # a task cannot depend on a task created in a for loop group since individual PipelineTask variables are reassigned after each loop iteration
                 dependent_group = group_name_to_group.get(
                     upstream_groups[0], None)
-                if isinstance(dependent_group,
-                              (tasks_group.Condition, tasks_group.ParallelFor)):
+                if isinstance(dependent_group, tasks_group.ParallelFor):
```
Comment (chensun) [inline sdk/python/kfp/compiler/compiler.py:631]: Shouldn't we add ExitHandler to the check instead of removing Condition from it?

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler.py:631]: Yes, I was thinking about this the wrong way

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler.py]: Added in [1c5a898](https://github.com/kubeflow/pipelines/pull/8088/commits/1c5a898dc38e439dbe3ab2b48d59fe433c707135).

Also added support for nested exit handlers to this PR with [ebee976](https://github.com/kubeflow/pipelines/pull/8088/commits/ebee976befe78c8650312625d2ff884ec0397193) and [ba5340c](https://github.com/kubeflow/pipelines/pull/8088/commits/ba5340cfe3d59cfb372d033781fad22be75ea360)

File: sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_nested_exit_handlers.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+def producer_op() -> str:
+    return 'a'
+
+
+@dsl.component
+def dummy_op(msg: str = '') -> str:
+    return msg
+
+
+@dsl.pipeline(name='pipeline-with-nested-exit-handlers')
```
Comment (chensun) [inline sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_nested_exit_handlers.py:28]: I would avoid this nested example unless 1) we have explicit ask for this feature, and 2) we confirmed there's a support path (via Argo) on KFP. 
Before we officially support this feature, we should ideally block it from compiling.

File: samples/test/config.yaml [hunk 4]
```diff
@@ -122,6 +122,8 @@
   path: samples.v2.producer_consumer_param_test
 - name: pipeline_with_importer
   path: samples.v2.pipeline_with_importer_test
+- name: pipeline_with_multiple_exit_handlers
+  path: samples.v2.pipeline_with_multiple_exit_handlers
```
[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply samples/test/config.yaml]: Removed

File: sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_nested_exit_handlers.py (hunk 3 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_nested_exit_handlers.py]: That makes sense to me. I left the logic, but added a compilation guard and a comment in debfe4b.

**Review by chensun — COMMENTED** [review]
/lgtm

Thanks!

Maybe add a release note for this change?

**Comment by chensun** [comment]
/lgtm
/approve

Thanks @connor-mccarthy !

**Pull request merged**